### DISCUSSION TITLE
Annotate Mutate() func with `InstantHandle`

### DISCUSSION THREAD

**Discussion by Socolin** [SixLabors/ImageSharp#2745] [Ideas]
Hello,

When writing code like

```
using (var img1 = Image.Load("cat-original.png"))
using (var img2 = Image.Load("cat-yellow.png"))
{
	img1.Mutate(c => c.DrawImage(img2, new Point(0, 0), new GraphicsOptions()));
}
```

On rider I'm getting a warning saying `Captured variable is disposed in the outer scope`
![image](https://github.com/SixLabors/ImageSharp/assets/539620/71b269e2-3c17-421f-980b-c333b0731187)

Since the lambda is executed immediatly I think this warning is a false positive, would it be possible to annotate the func with `[InstanHandle]` (https://www.jetbrains.com/help/rider/Reference__Code_Annotation_Attributes.html#PureAttribute) ?

**Comment by antonfirsov** [comment]
Would be nice if such an attribute was part of the BCL. ImageSharp is dependency-free by design and taking a dependency on `JetBrains.Annotations` would break that.

**Reply by Socolin** [reply]
So I asked to Jetbrains and it seems it's possible to annotate from another NuGet https://www.jetbrains.com/help/resharper/Code_Analysis__External_Annotations.html#creating could another nuget like `ImageSharp.JetbrainsAnnotation` be an option ?

**Reply by citizenmatt** [reply]
It's also possible to include the annotations in source form. The attributes are marked as `internal`, so they're not exported as an API, but Rider/ReSharper's code analysis will still see them and consumers of your library will benefit. [Web help here](https://www.jetbrains.com/help/resharper/Code_Analysis__Annotations_in_Source_Code.html#embedding-declarations-of-code-annotations-in-your-source-code) and [here's a very old blog post that's still valid](https://blog.jetbrains.com/dotnet/2015/08/12/how-to-use-jetbrains-annotations-to-improve-resharper-inspections/#including-the-source).

**Reply by Socolin** [reply]
So I created the XML for my project for now, I can complete it if needed, but if it's possible to annotate it from the code it would be easier I think

Here the xml
```xml
<assembly name="SixLabors.ImageSharp">
  <member name="M:SixLabors.ImageSharp.Processing.ProcessingExtensions.Mutate(SixLabors.ImageSharp.Image,System.Action{SixLabors.ImageSharp.Processing.IImageProcessingContext})">
    <parameter name="operation">
      <attribute ctor="M:JetBrains.Annotations.InstantHandleAttribute.#ctor"/>
    </parameter>
  </member>
  <member name="M:SixLabors.ImageSharp.Processing.ProcessingExtensions.Mutate``1(SixLabors.ImageSharp.Image{``0},System.Action{SixLabors.ImageSharp.Processing.IImageProcessingContext})">
    <parameter name="operation">
      <attribute ctor="M:JetBrains.Annotations.InstantHandleAttribute.#ctor"/>
    </parameter>
  </member>
</assembly>
```

**Reply by JimBobSquarePants** [reply]
I'm concerned that we're hacking around an issue with the analyzer. 

1. If the analyzer requires updating, can it be fixed at the source?
2. If it cannot be updated to avoid false positives, is it even useful?

**Reply by Socolin** [reply]
It's built-in analyzer in R# / Rider.

It detect if you give a disposable to a Function/Action that may be used later, so the analyzer is useful for that, but currently it cannot determine if the action is going to be called immediately or not, maybe in the present case of `.Mutate` it could be determined automatically, but exploring the code path to determine this might be expensive, and there is some scenario (when the implementation is not known) where it cannot determine when the lambda is called.

To avoid the false positive, Jetbrains is providing a bunch of annotation in a nuget that can be used to remove false positive from analyzers.

That way, instead of having no analyzer at all on some topic, because some case might be false positive, the analyzer is here with way to remove those false positive.

I would also prefer if Microsoft add a bunch of annotation to help with those scenario, but as long as VS is not going to support this they are not going to add those kind of annotation I think.

I understand you don't want to add Jetbrains specific stuff in this project, and there is a way to provide the annotation on a separate NuGet or directly in the project with the XML

I'm happy with maintaining the annotation on my side for my project, it would be easier to maintain those if it can be done directly in the code and fix those false positive to everyone using R# / Rider.
